url.go

```diff
@@ -152,64 +157,111 @@ func ParseReference(url string) (*Reference, error) {
 
 func parseReference(url string) (*Reference, error) {
 	var r Reference
-	i := strings.Index(url, ":")
-	if i >= 0 {
-		r.Schema = url[:i]
-		if r.Schema != "cs" && r.Schema != "local" {
-			return nil, fmt.Errorf("entity URL has invalid schema: %q", url)
+	// Check if we're dealing with a v1 or v2 URL.
+	// TODO: this is a naive check that only works on fully-qualified URLs;
+	// further work will be required for future URLs
+	// XXX Get rid of urlVersion, test based on validStore.MatchString
+	if validStore.MatchString(url) {
+		r.Schema = "cs"
+		url = validStore.ReplaceAllString(url, "")
+		parts := strings.Split(strings.Trim(url, "/"), "/")
+		if len(parts) == 0 {
+			return nil, fmt.Errorf("entity URL malformed, no parts: %q", url)
+		}
+		if parts[0] == "u" {
+			if len(parts) < 3 {
+				return nil, fmt.Errorf("entity URL malformed, expecting user and name: %q", url)
+			}
+			r.User = parts[1]
+			parts = parts[2:]
+		}
+		r.Name = parts[0]
+		parts = parts[1:]
+		r.Revision = -1
+		if len(parts) > 0 {
```

Comment:
If this was a separate function, you could do:

   if len(parts) == 0 {
       return &r, nil
    }